drivers/platform/xilinx/xilinx_spi_pl.c

```diff
@@ -0,0 +1,301 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <xparameters.h>
+
+#ifdef XPAR_XSPI_NUM_INSTANCES
+
+#include <xspi.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "spi.h"
+#include "error.h"
+#include "delay.h"
+#include "util.h"
+
+#define XSP_DEFAULT_SR_VALUE	0x25
+
+#if XPAR_SPI_0_FIFO_EXIST==1
+#define FIFO_DEPTH XPAR_SPI_0_FIFO_DEPTH
+#else
+#define FIFO_DEPTH 1
+#endif
+
+struct xspi_desc {
+	uint32_t base_addr;
+	uint32_t disable_slaves_mask;
+};
+
+/*
+ * Used to store current spi device configuration in order to not update it for each spi
+ * transaction if not needed.
+ */
+static struct spi_desc g_spi_desc;
+
+/* Read device register */
+static inline uint32_t _read(struct xspi_desc *xdesc, uint32_t reg)
+{
+	return Xil_In32(xdesc->base_addr + reg);
+}
+
+/* Write device register */
+static inline void _write(struct xspi_desc *xdesc, uint32_t reg, uint32_t val)
+{
+	Xil_Out32(xdesc->base_addr + reg, val);
+}
+
+/* Update device register. Only bits set to 1 in mask will be modified */
+static inline void _update(struct xspi_desc *xdesc, uint32_t reg,
+				uint32_t val, uint32_t mask)
+{
+	uint32_t old;
+
+	old = _read(xdesc, reg);
+	old &= ~mask;
+	old |= val & mask;
+	_write(xdesc, reg, old);
+}
+
+/* Update spi device configuration if needed */
+static void _update_mode(struct spi_desc *desc)
+{
+	uint32_t		val;
+	uint32_t		mask;
+
+	if (desc->mode == g_spi_desc.mode &&
+			desc->bit_order == g_spi_desc.bit_order)
+		return ;
+
+	/* Update mode */
+	val = desc->mode | SPI_CPOL ? XSP_CR_CLK_POLARITY_MASK : 0;
+	val |= desc->mode | SPI_CPHA ? XSP_CR_CLK_PHASE_MASK : 0;
+	mask = XSP_CR_CLK_POLARITY_MASK | XSP_CR_CLK_PHASE_MASK;
+
+	val |= desc->bit_order == SPI_BIT_ORDER_LSB_FIRST ?
+			XSP_CR_LSB_MSB_FIRST_MASK : 0;
+	mask |= XSP_CR_LSB_MSB_FIRST_MASK;
+
+	_update(desc->extra, XSP_CR_OFFSET, val, mask);
+
+	g_spi_desc = *desc;
+}
+
+/* Configure spi */
+static int32_t _xil_spi_init(struct xspi_desc *xdesc)
+{
+	uint32_t val;
+
+	/* Software reset */
+	_write(xdesc, XSP_SRR_OFFSET, XSP_SRR_RESET_MASK);
+	mdelay(1);
+
+	/* Check if success reset */
+	val = _read(xdesc, XSP_SR_OFFSET);
+	if (val != XSP_DEFAULT_SR_VALUE)
+		return -ENODEV;
+
+	/* Disable interrupts. Drivers only works with pooled mode */
+	_write(xdesc, XSP_IIER_OFFSET, 0);
+
+	/* Deasert chip selects */
+	_write(xdesc, XSP_SSR_OFFSET, 0xFFFFFFFF);
+
+	/* Configure spi device */
+	val = XSP_CR_ENABLE_MASK |
+		/* Maybe not needed */
+		//XSP_CR_MANUAL_SS_MASK |
+		XSP_CR_MASTER_MODE_MASK |
+		XSP_CR_TRANS_INHIBIT_MASK;
+
+	_write(xdesc, XSP_CR_OFFSET, val);
+
+	return SUCCESS;
+}
+
+static int32_t xil_spi_init_pl(struct spi_desc **desc,
+	const struct spi_init_param *param)
+{
+	struct spi_desc		*ldesc;
+	struct xspi_desc	*xdesc;
+	int32_t			err;
+	XSpi_Config		*cfg;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	err = -EINVAL;
```

Comment:
This is not a good way to do it. 
`err` should be set only when you encounter an error like below
```
if (!cfg) {
    err = ...
    goto error;
}

I agree it does the job, but It's awkard to read as it is.

Reply:
Lazy me wanted to have less code lines. Changed anyway.